inject/titus-inject-metadataproxy.c

```diff
@@ -126,7 +126,14 @@ static void setup_veth(int private_ns_fd, int container_ns_fd) {
 	rtnl_link_set_name(peer, IN_CONTAINER_INTERFACE_NAME);
 	rtnl_link_set_ns_fd(peer, container_ns_fd);
 
-	err = rtnl_link_add(nls, veth, NLM_F_CREATE | NLM_F_EXCL);
+	for (int i = 0; i < 5; i++) {
+		err = rtnl_link_add(nls, veth, NLM_F_CREATE | NLM_F_EXCL);
```

Comment:
Can you dump the values of nls / veth? Is there an rtnl pretty print function?

Reply:
Just kinda doing my own research, never used libnl before.
Looking at the call graph of this function:
https://www.infradead.org/~tgr/libnl/doc/api/group__link.html#ga010cb3c6c598f7a60f97b67b8df9b227
I think the only place we could get his error is here when we send our request over the netlink socket:
https://www.infradead.org/~tgr/libnl/doc/api/nl_8c_source.html#l00251

Not unexpectedly, I think this just means it comes from the kernel (of course, since the thing that changed was the new kernel version).

From here I don't know how to trace into the kernel (`RTM_NEWLINK`?) what is going on.

If we could figure out what can cause perm denied in the kernel when we make new links like this, it may be able to answer the question of "will retries even help?"

Still though, I'm in favor of trying anything. Let's see if they help :)

Reply:
You're correct that it's coming from the kernel, though my attempts to catch it via bpftrace haven't turned anything up (it's not `rtnl_newlink()` returing EPERM, which is unexpected).  My theory is that it's a race, so if retries make the problem go away, then we know for sure.

There's a pretty print function - it didn't seem to include much in the way of useful information, but I can add it just in case.